File: 2/2000s/23s/2300/SuccessfulPairsOfSpellsAndPotions.cpp
/*
https://leetcode.com/problems/successful-pairs-of-spells-and-potions/description/?envType=daily-question&envId=2025-10-08
*/

class Solution_TLE {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        vector<int>res;
        int sz1=spells.size(), sz2=potions.size();
        for(int i=0; i<sz1; ++i){
            int curSpell=spells[i], curCount=0;
            for(int j=0; j<sz2; ++j){
                long long curProd=(long long)curSpell*potions[j];
                if(curProd>=success){
                    curCount++;
                }
            }
            res.push_back(curCount);
        }
        return res;
    }
};

// ------- ******* -------

class Solution_OPT {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        sort(potions.begin(), potions.end());
        vector<int>res;
        int sz=potions.size();
        for(int spell: spells){
            long long req=(success+spell-1)/spell;
            int idx=lower_bound(potions.begin(), potions.end(), req)-potions.begin();
            res.push_back(sz-idx);
        }
        return res;
    }
};

File: 2/2000s/23s/2300/SuccessfulPairsOfSpellsAndPotions.java
/*
https://leetcode.com/problems/successful-pairs-of-spells-and-potions/description/?envType=daily-question&envId=2025-10-08
*/

class Solution {

    public static boolean isValid(int currentPortion, long minPotionRequired){
        return currentPortion>=minPotionRequired;
    }

    public static int binarySearch(int[] portions, long minPotionRequired){
        int left=0, right=portions.length-1;
        int resultantIndex=right+1;
        while(left<=right){
            int mid=left+(right-left)/2;
            if(isValid(portions[mid], minPotionRequired)){
                resultantIndex=mid;
                right=mid-1;
            } else left=mid+1;
        }
        return resultantIndex;
    }

    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        int lengthOfSpells=spells.length, lengthOfPotions=potions.length;
        int[] res=new int[lengthOfSpells];
        for(int i=0; i<lengthOfSpells; ++i){
            int currentSpell=spells[i];
            long minPotionRequired=(success+currentSpell-1)/currentSpell;
            int firstValidIndex=binarySearch(potions, minPotionRequired);
            res[i]=(firstValidIndex==lengthOfPotions)?0:(lengthOfPotions-firstValidIndex);
        }
        return res;
    }
}